dynamodb/schema.go

```diff
@@ -0,0 +1,358 @@
+// rows to infer the schema for optional attributes.
+func ProcessSchema(conv *internal.Conv, client dynamoClient, tables []string, sampleSize int64) error {
+	if len(tables) == 0 {
+		var err error
+		tables, err = listTables(client)
+		if err != nil {
+			return err
+		}
+		if len(tables) == 0 {
+			return fmt.Errorf("no DynamoDB table exists under this account")
+		}
+	}
+	for _, t := range tables {
+		if err := processTable(conv, client, t, sampleSize); err != nil {
+			return err
+		}
+	}
+	schemaToDDL(conv)
+	conv.AddPrimaryKeys()
+	return nil
+}
+
+func listTables(client dynamoClient) ([]string, error) {
+	var tables []string
+	var lastEvaluatedTableName *string
+
+	for {
+		input := &dynamodb.ListTablesInput{}
+		if lastEvaluatedTableName != nil {
+			input.ExclusiveStartTableName = lastEvaluatedTableName
+		}
+
+		result, err := client.ListTables(input)
+		if err != nil {
+			return nil, err
+		}
+		for _, t := range result.TableNames {
+			tables = append(tables, *t)
+		}
+
+		if result.LastEvaluatedTableName == nil {
+			return tables, nil
+		}
+		lastEvaluatedTableName = result.LastEvaluatedTableName
+	}
+}
+
+func processTable(conv *internal.Conv, client dynamoClient, table string, sampleSize int64) error {
+	dySchema := dynamoDBSchema{TableName: table}
+	err := dySchema.parseIndexes(client)
+	if err != nil {
+		return err
+	}
+	stats, count, err := dySchema.scanSampleData(client, sampleSize)
+	if err != nil {
+		return err
+	}
+	dySchema.inferDataTypes(stats, count)
+	conv.SrcSchema[table] = dySchema.genericSchema()
+	return nil
+}
+
+type dynamoDBSchema struct {
+	TableName   string
+	ColumnNames []string
+	ColumnTypes map[string]colType
+	PrimaryKeys []string
+	SecIndexes  []index
+}
+
+type colType struct {
+	Name     string
+	Nullable bool
+}
+
+type index struct {
+	Name string
+	Keys []string
+}
+
+func (s *dynamoDBSchema) parseIndexes(client dynamoClient) error {
+	input := &dynamodb.DescribeTableInput{
+		TableName: aws.String(s.TableName),
+	}
+
+	result, err := client.DescribeTable(input)
+	if err != nil {
+		return fmt.Errorf("failed to make a DescribeTable API call for table %v: %v", s.TableName, err)
+	}
+
+	// Primary keys
+	for _, i := range result.Table.KeySchema {
+		s.PrimaryKeys = append(s.PrimaryKeys, *i.AttributeName)
+	}
+
+	// Secondary indexes
+	for _, i := range result.Table.GlobalSecondaryIndexes {
+		var keys []string
+		for _, j := range i.KeySchema {
+			keys = append(keys, *j.AttributeName)
+		}
+		s.SecIndexes = append(s.SecIndexes, index{Name: *i.IndexName, Keys: keys})
+	}
+
+	return nil
+}
+
+func (s *dynamoDBSchema) scanSampleData(client dynamoClient, sampleSize int64) (map[string]map[string]int64, int64, error) {
+	// A map from column name to a count map of possible data types.
+	stats := make(map[string]map[string]int64)
+	var lastEvaluatedKey map[string]*dynamodb.AttributeValue
+	var count int64
+
+	for {
+		// Build the query input parameters
+		params := &dynamodb.ScanInput{
```

Comment:
Do we need to create dynamodb.ScanInput for each iteration of the loop?
Could we just define one before the start of loop and re-use it.
And then get rid of the lastEvaluatedKey var (and line 175 to 177) and instead change line 203 to:

params.ExclusiveStartKey = result.LastEvaluatedKey

I think that would be less code and a bit easier to understand.

Reply:
Nice suggestion. Fixed.